Add optional numpy parallel compiler, fix some syntax and add Python 3.11 support

chris-allan opened the pull request:
Adds the numpy parallel compiler as an option. This is deprecated upstream [^1] but is a very nice and quick way to accelerate these builds. Python 3.11 support has also been added. See also:

* zeroc-ice/ice#1393
* zeroc-ice/ice#1394
* ome/omero-py#360

[^1]: https://numpy.org/devdocs/reference/distutils_status_migration.html

chris-allan committed:
Add optional numpy parallel compiler and fix some setup.py syntax

chris-allan committed:
Use 3 jobs

chris-allan committed:
Upgrade actions to latest version

sbesson reviewed (commented):
Successfully tested a minimal installation, login & query workflow using various Python version installed using https://github.com/pyenv/pyenv 
- Python 3.8

```
sbesson@Sebastiens-MacBook-Pro-3 /tmp % ~/.pyenv/versions/3.8.18/bin/python -m venv venv
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/pip install ~/Downloads/artifacts/zeroc_ice-3.6.5-cp38-cp38-macosx_11_0_x86_64.whl 
Processing /Users/sbesson/Downloads/artifacts/zeroc_ice-3.6.5-cp38-cp38-macosx_11_0_x86_64.whl
Installing collected packages: zeroc-ice
Successfully installed zeroc-ice-3.6.5

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /private/tmp/venv/bin/python -m pip install --upgrade pip
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/pip install omero-py
Collecting omero-py
  Using cached omero_py-5.16.1-py3-none-any.whl (2.8 MB)
Collecting portalocker
  Using cached portalocker-2.8.2-py3-none-any.whl (17 kB)
Collecting numpy
  Downloading numpy-1.24.4-cp38-cp38-macosx_10_9_x86_64.whl (19.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.8/19.8 MB 26.2 MB/s eta 0:00:00
Collecting appdirs
  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting urllib3<2
  Using cached urllib3-1.26.17-py2.py3-none-any.whl (143 kB)
Collecting future
  Using cached future-0.18.3.tar.gz (840 kB)
  Preparing metadata (setup.py) ... done
Collecting PyYAML
  Downloading PyYAML-6.0.1-cp38-cp38-macosx_10_9_x86_64.whl (191 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 191.7/191.7 kB 6.2 MB/s eta 0:00:00
Collecting requests
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Requirement already satisfied: zeroc-ice<3.7,>=3.6.5 in ./venv/lib/python3.8/site-packages (from omero-py) (3.6.5)
Collecting Pillow
  Downloading Pillow-10.0.1-cp38-cp38-macosx_10_10_x86_64.whl (3.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 17.2 MB/s eta 0:00:00
Collecting certifi>=2017.4.17
  Using cached certifi-2023.7.22-py3-none-any.whl (158 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Collecting charset-normalizer<4,>=2
  Downloading charset_normalizer-3.3.0-cp38-cp38-macosx_10_9_x86_64.whl (118 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.3/118.3 kB 4.5 MB/s eta 0:00:00
Installing collected packages: appdirs, urllib3, PyYAML, portalocker, Pillow, numpy, idna, future, charset-normalizer, certifi, requests, omero-py
  DEPRECATION: future is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559
  Running setup.py install for future ... done
Successfully installed Pillow-10.0.1 PyYAML-6.0.1 appdirs-1.4.4 certifi-2023.7.22 charset-normalizer-3.3.0 future-0.18.3 idna-3.4 numpy-1.24.4 omero-py-5.16.1 portalocker-2.8.2 requests-2.31.0 urllib3-1.26.17

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /private/tmp/venv/bin/python -m pip install --upgrade pip
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero login public@idr.openmicroscopy.org
Password:
Created session for public@idr.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Public
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero hql 'select count(p) from Project p' -q
 # | Col1 
---+------
 0 | 122  
(1 row)
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero logout
sbesson@Sebastiens-MacBook-Pro-3 /tmp % rm -rf venv 
```

- Python 3.9

```
sbesson@Sebastiens-MacBook-Pro-3 /tmp %  ~/.pyenv/versions/3.9.18/bin/python -m venv venv
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/pip install ~/Downloads/artifacts/zeroc_ice-3.6.5-cp39-cp39-macosx_10_9_x86_64.whl 
Processing /Users/sbesson/Downloads/artifacts/zeroc_ice-3.6.5-cp39-cp39-macosx_10_9_x86_64.whl
Installing collected packages: zeroc-ice
Successfully installed zeroc-ice-3.6.5

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /private/tmp/venv/bin/python -m pip install --upgrade pip
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/pip install omero-py
Collecting omero-py
  Using cached omero_py-5.16.1-py3-none-any.whl (2.8 MB)
Collecting numpy
  Using cached numpy-1.26.0-cp39-cp39-macosx_10_9_x86_64.whl (20.6 MB)
Collecting PyYAML
  Using cached PyYAML-6.0.1-cp39-cp39-macosx_10_9_x86_64.whl (197 kB)
Collecting urllib3<2
  Using cached urllib3-1.26.17-py2.py3-none-any.whl (143 kB)
Collecting appdirs
  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting future
  Using cached future-0.18.3-py3-none-any.whl
Collecting requests
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Collecting Pillow
  Using cached Pillow-10.0.1-cp39-cp39-macosx_10_10_x86_64.whl (3.7 MB)
Requirement already satisfied: zeroc-ice<3.7,>=3.6.5 in ./venv/lib/python3.9/site-packages (from omero-py) (3.6.5)
Collecting portalocker
  Using cached portalocker-2.8.2-py3-none-any.whl (17 kB)
Collecting charset-normalizer<4,>=2
  Using cached charset_normalizer-3.3.0-cp39-cp39-macosx_10_9_x86_64.whl (119 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2023.7.22-py3-none-any.whl (158 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Installing collected packages: appdirs, urllib3, PyYAML, portalocker, Pillow, numpy, idna, future, charset-normalizer, certifi, requests, omero-py
Successfully installed Pillow-10.0.1 PyYAML-6.0.1 appdirs-1.4.4 certifi-2023.7.22 charset-normalizer-3.3.0 future-0.18.3 idna-3.4 numpy-1.26.0 omero-py-5.16.1 portalocker-2.8.2 requests-2.31.0 urllib3-1.26.17

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /private/tmp/venv/bin/python -m pip install --upgrade pip
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero login public@idr.openmicroscopy.org                                         
Previously logged in to idr.openmicroscopy.org:4064 as public
Password:
Created session for public@idr.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Public
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero hql 'select count(p) from Project p' -q                                     
 # | Col1 
---+------
 0 | 122  
(1 row)
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero logout
sbesson@Sebastiens-MacBook-Pro-3 /tmp % rm -r venv 
```

- Python 3.10

```
sbesson@Sebastiens-MacBook-Pro-3 /tmp % ~/.pyenv/versions/3.10.13/bin/python -m venv venv
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/pip install ~/Downloads/artifacts/zeroc_ice-3.6.5-cp310-cp310-macosx_11_0_x86_64.whl 
Processing /Users/sbesson/Downloads/artifacts/zeroc_ice-3.6.5-cp310-cp310-macosx_11_0_x86_64.whl
Installing collected packages: zeroc-ice
Successfully installed zeroc-ice-3.6.5

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /private/tmp/venv/bin/python -m pip install --upgrade pip
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/pip install omero-py
Collecting omero-py
  Using cached omero_py-5.16.1-py3-none-any.whl (2.8 MB)
Collecting urllib3<2
  Using cached urllib3-1.26.17-py2.py3-none-any.whl (143 kB)
Collecting PyYAML
  Downloading PyYAML-6.0.1-cp310-cp310-macosx_10_9_x86_64.whl (189 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 189.4/189.4 kB 2.8 MB/s eta 0:00:00
Collecting Pillow
  Downloading Pillow-10.0.1-cp310-cp310-macosx_10_10_x86_64.whl (3.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 12.7 MB/s eta 0:00:00
Collecting portalocker
  Using cached portalocker-2.8.2-py3-none-any.whl (17 kB)
Collecting appdirs
  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting numpy
  Downloading numpy-1.26.0-cp310-cp310-macosx_10_9_x86_64.whl (20.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.6/20.6 MB 27.7 MB/s eta 0:00:00
Collecting requests
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Collecting future
  Using cached future-0.18.3-py3-none-any.whl
Requirement already satisfied: zeroc-ice<3.7,>=3.6.5 in ./venv/lib/python3.10/site-packages (from omero-py) (3.6.5)
Collecting idna<4,>=2.5
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2023.7.22-py3-none-any.whl (158 kB)
Collecting charset-normalizer<4,>=2
  Downloading charset_normalizer-3.3.0-cp310-cp310-macosx_10_9_x86_64.whl (119 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.8/119.8 kB 4.3 MB/s eta 0:00:00
Installing collected packages: appdirs, urllib3, PyYAML, portalocker, Pillow, numpy, idna, future, charset-normalizer, certifi, requests, omero-py
Successfully installed Pillow-10.0.1 PyYAML-6.0.1 appdirs-1.4.4 certifi-2023.7.22 charset-normalizer-3.3.0 future-0.18.3 idna-3.4 numpy-1.26.0 omero-py-5.16.1 portalocker-2.8.2 requests-2.31.0 urllib3-1.26.17

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /private/tmp/venv/bin/python -m pip install --upgrade pip
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero login public@idr.openmicroscopy.org                                           
Previously logged in to idr.openmicroscopy.org:4064 as public
Password:
Created session for public@idr.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Public
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero hql 'select count(p) from Project p' -q                                       
 # | Col1 
---+------
 0 | 122  
(1 row)
sbesson@Sebastiens-MacBook-Pro-3 /tmp % venv/bin/omero logout
```

Should we add Python 3.11 and possibly 3.12 to this list so that we cover all versions above 3.8?

chris-allan committed:
Explicitly set the deployment target to macOS 10.13

chris-allan committed:
Update deployment target to macOS 11.0

chris-allan committed:
Add Python 3.11 to the build matrix

chris-allan committed:
Add support for Python 3.11 (See zeroc-ice/ice#1394)

This patch is nearly identical to
zeroc-ice/ice@4d1158b57b20081aeafc0613240fb2f27ac323ed with the
exception that I have just removed the compile.h and eval.h header
inclusions.  They have not been required since Python 2.4 as they were
moved into Python.h.

sbesson reviewed (approved):
Tested on OS X 13.6 (Intel) with the following Python:
- the system Python 3.9 installed under `/usr/bin/python3` as part of the XCode Command Line Tools
- Python 3.8, 3.9, 3.10, 3.11 installed using [pyenv](https://github.com/pyenv/pyenv)
- Python 3.8, 3.9, 3.10, 3.11 installed using [Homebrew](https://brew.sh/)
- Python 3.8, 3.9, 3.10, 3.11 installed using [Miniconda](https://docs.conda.io/en/latest/miniconda.html/)

For each Python version, a new virtual environment was created and the corresponding wheel was installed.
All wheels were successfully installed except for the Conda environments which all failed with

```
(base) sbesson@Sebastiens-MacBook-Pro-3 Downloads % ./venv_pyenv_311/bin/pip install artifacts/zeroc_ice-3.6.5-cp311-cp311-macosx_11_0_universal2.whl 
Processing ./artifacts/zeroc_ice-3.6.5-cp311-cp311-macosx_11_0_universal2.whl
Installing collected packages: zeroc-ice
Successfully installed zeroc-ice-3.6.5
(base) sbesson@Sebastiens-MacBook-Pro-3 Downloads % ./venv_conda_38/bin/pip install artifacts/zeroc_ice-3.6.5-cp38-cp38-macosx_11_0_x86_64.whl 
ERROR: zeroc_ice-3.6.5-cp38-cp38-macosx_11_0_x86_64.whl is not a supported wheel on this platform.

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /Users/sbesson/Downloads/venv_conda_38/bin/python -m pip install --upgrade pip
(base) sbesson@Sebastiens-MacBook-Pro-3 Downloads % ./venv_conda_39/bin/pip install artifacts/zeroc_ice-3.6.5-cp39-cp39-macosx_11_0_x86_64.whl 
ERROR: zeroc_ice-3.6.5-cp39-cp39-macosx_11_0_x86_64.whl is not a supported wheel on this platform.

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /Users/sbesson/Downloads/venv_conda_39/bin/python -m pip install --upgrade pip
(base) sbesson@Sebastiens-MacBook-Pro-3 Downloads % ./venv_conda_310/bin/pip install artifacts/zeroc_ice-3.6.5-cp310-cp310-macosx_11_0_x86_64.whl 
ERROR: zeroc_ice-3.6.5-cp310-cp310-macosx_11_0_x86_64.whl is not a supported wheel on this platform.

[notice] A new release of pip is available: 23.0.1 -> 23.2.1
[notice] To update, run: /Users/sbesson/Downloads/venv_conda_310/bin/python -m pip install --upgrade pip
(base) sbesson@Sebastiens-MacBook-Pro-3 Downloads % ./venv_conda_311/bin/pip install artifacts/zeroc_ice-3.6.5-cp311-cp311-macosx_11_0_universal2.whl 
ERROR: zeroc_ice-3.6.5-cp311-cp311-macosx_11_0_universal2.whl is not a supported wheel on this platform.
```

Note also https://docs.conda.io/projects/conda-build/en/latest/user-guide/wheel-files.html which does not recommend using wheels within Conda environment as the canonical way.

For all the other environments above, `omero-py` was installed and a minimal connection + query was tested against `idr.openmicroscopy.org`. Everything worked as expected except for the Python 3.11 environment which failed with a legitimate failure as one of the modules is using a dropped feature. 

From my perspective, this is definitely a step forward and I see no objection to merging the changes as they stand. I will capture into the Python 3.11/OMERO.py compatibility problems as an upstream issue
